File: Level 3 - Advanced Problem Solving/treap/core_problems.cpp
#include <iostream>
#include <random>
#include <ctime>

/**
 * Core Treap Problems in C++
 * Focuses on set size and order statistics.
 */

struct Node {
    int key, priority, size;
    Node *left, *right;

    Node(int k) : key(k), priority(std::rand()), size(1), left(nullptr), right(nullptr) {}

    void updateSize() {
        size = 1 + (left ? left->size : 0) + (right ? right->size : 0);
    }
};

void split(Node* root, int target, Node*& L, Node*& R) {
    if (!root) { L = R = nullptr; return; }
    if (root->key <= target) {
        split(root->right, target, root->right, R);
        L = root;
    } else {
        split(root->left, target, L, root->left);
        R = root;
    }
    if (root) root->updateSize();
}

void merge(Node*& root, Node* L, Node* R) {
    if (!L || !R) { root = L ? L : R; return; }
    if (L->priority > R->priority) {
        merge(L->right, L->right, R);
        root = L;
    } else {
        merge(R->left, L, R->left);
        root = R;
    }
    if (root) root->updateSize();
}

int countLessEqual(Node*& root, int X) {
    Node *L, *R;
    split(root, X, L, R);
    int res = (L ? L->size : 0);
    merge(root, L, R);
    return res;
}

int findKth(Node* root, int k) {
    int leftSize = (root->left ? root->left->size : 0);
    if (k == leftSize + 1) return root->key;
    if (k <= leftSize) return findKth(root->left, k);
    return findKth(root->right, k - leftSize - 1);
}

int main() {
    std::srand(std::time(0));
    Node* root = nullptr;
    int data[] = {5, 12, 7, 20, 15, 3, 10};

    for (int x : data) {
        Node *L, *R;
        split(root, x, L, R);
        Node* newNode = new Node(x);
        Node* temp;
        merge(temp, L, newNode);
        merge(root, temp, R);
    }

    std::cout << "Current Set Size: " << (root ? root->size : 0) << std::endl;
    std::cout << "Elements <= 12: " << countLessEqual(root, 12) << std::endl;
    std::cout << "2nd smallest element: " << findKth(root, 2) << std::endl;
    std::cout << "5th smallest element: " << findKth(root, 5) << std::endl;

    return 0;
}

File: Level 3 - Advanced Problem Solving/treap/core_problems.java
import java.util.Random;

/**
 * Core Treap Problems in Java
 * Focuses on set-like operations and range count queries.
 */
public class core_problems {
    private static final Random rand = new Random();

    static class Node {
        int key, priority, size;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.priority = rand.nextInt();
            this.size = 1;
        }

        void updateSize() {
            this.size = 1 + (left != null ? left.size : 0) + (right != null ? right.size : 0);
        }
    }

    static class Pair {
        Node first, second;

        Pair(Node first, Node second) {
            this.first = first;
            this.second = second;
        }
    }

    public static Pair split(Node root, int target) {
        if (root == null)
            return new Pair(null, null);
        if (root.key <= target) {
            Pair p = split(root.right, target);
            root.right = p.first;
            root.updateSize();
            return new Pair(root, p.second);
        } else {
            Pair p = split(root.left, target);
            root.left = p.second;
            root.updateSize();
            return new Pair(p.first, root);
        }
    }

    public static Node merge(Node L, Node R) {
        if (L == null)
            return R;
        if (R == null)
            return L;
        if (L.priority > R.priority) {
            L.right = merge(L.right, R);
            L.updateSize();
            return L;
        } else {
            R.left = merge(L, R.left);
            R.updateSize();
            return R;
        }
    }

    // Problem 1: Count elements <= X
    public static int countLessEqual(Node root, int X) {
        Pair p = split(root, X);
        int count = (p.first != null ? p.first.size : 0);
        root = merge(p.first, p.second); // Reconstruct
        return count;
    }

    // Problem 2: Find K-th smallest element (1-indexed)
    public static int findKth(Node root, int k) {
        int leftSize = (root.left != null ? root.left.size : 0);
        if (k == leftSize + 1)
            return root.key;
        if (k <= leftSize)
            return findKth(root.left, k);
        return findKth(root.right, k - leftSize - 1);
    }

    public static void main(String[] args) {
        Node root = null;
        int[] data = { 5, 12, 7, 20, 15, 3, 10 };
        for (int x : data) {
            Pair p = split(root, x);
            root = merge(merge(p.first, new Node(x)), p.second);
        }

        System.out.println("Current Set Size: " + (root != null ? root.size : 0));
        System.out.println("Elements <= 12: " + countLessEqual(root, 12));
        System.out.println("2nd smallest element: " + findKth(root, 2));
        System.out.println("5th smallest element: " + findKth(root, 5));
    }
}
